Source/WebCore/css/CSSSelector.cpp

```diff
@@ -358,22 +359,32 @@ std::optional<CSSSelector::PseudoElement> CSSSelector::parsePseudoElement(String
     return type;
 }
 
-std::optional<PseudoId> CSSSelector::parseStandalonePseudoElement(StringView input, const CSSSelectorParserContext& context)
+// FIXME: We should eventually deduplicate this with CSSSelectorParser::consumePseudo() somehow.
+std::optional<PseudoId> CSSSelector::parseStandalonePseudoElement(const String& input, const CSSSelectorParserContext& context)
 {
-    // FIXME: Tokenize input.
-    if (input.startsWith("::"_s)) {
-        auto pseudoElement = parsePseudoElement(input.substring(2), context);
-        if (!pseudoElement)
-            return std::nullopt;
-        return pseudoId(*pseudoElement);
-    }
-    if (input.startsWith(":"_s)) {
-        auto pseudoClassOrElement = parsePseudoClassAndCompatibilityElementString(input.substring(1));
+    // FIXME: Add support for FunctionToken (webkit.org/b/264103).
+    auto tokenizer = CSSTokenizer { input };
+    auto range = tokenizer.tokenRange();
+    if (range.peek().type() != ColonToken)
+        return std::nullopt;
+    range.consume();
+    auto token = range.peek();
+    if (token.type() == IdentToken) {
+        auto pseudoClassOrElement = parsePseudoClassAndCompatibilityElementString(token.value());
         if (!pseudoClassOrElement.compatibilityPseudoElement)
             return std::nullopt;
         return pseudoId(*pseudoClassOrElement.compatibilityPseudoElement);
     }
-    return std::nullopt;
+    if (token.type() != ColonToken)
+        return std::nullopt;
+    range.consume();
+    token = range.peek();
+    if (token.type() != IdentToken)
+        return std::nullopt;
+    auto pseudoElement = parsePseudoElement(token.value(), context);
+    if (!pseudoElement)
+        return std::nullopt;
```

Comment:
Should you check the end of the range here? I suppose you want to properly reject `::foo bar`

Reply:
(This should probably be tested fwiw)